packages/eslint-plugin/src/rules/no-misused-spread.ts

```diff
@@ -0,0 +1,189 @@
+  },
+];
+
+type MessageIds =
+  | 'noStringSpreadInArray'
+  | 'noSpreadInObject'
+  | 'noFunctionSpreadInObject'
+  | 'noClassInstanceSpreadInObject'
+  | 'noClassDeclarationSpreadInObject';
+
+export default createRule<Options, MessageIds>({
+  name: 'no-misused-spread',
+  meta: {
+    type: 'problem',
+    docs: {
+      description:
+        'Disallow using the spread operator when it might cause unexpected behavior',
+      recommended: 'strict',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      noStringSpreadInArray:
+        "Using the spread operator on a string can cause unexpected behavior. Prefer `String.split('')` instead.",
+
+      noSpreadInObject:
+        'Using the spread operator on `{{type}}` in an object can cause unexpected behavior.',
+
+      noFunctionSpreadInObject:
+        'Using the spread operator on a function without additional properties can cause unexpected behavior. Did you forget to call the function?',
+
+      noClassInstanceSpreadInObject:
+        'Using the spread operator on class instances without `[Symbol.iterator]` can cause unexpected behavior.',
+
+      noClassDeclarationSpreadInObject:
+        'Using the spread operator on class declarations can cause unexpected behavior. Did you forget to instantiate the class?',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          allowClassInstances: {
+            description:
+              'Whether to allow spreading class instances in objects.',
+            type: 'boolean',
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+  },
+
+  defaultOptions: [
+    {
+      allowClassInstances: false,
+    },
+  ],
+
+  create(context, [options]) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+
+    function checkArraySpread(node: TSESTree.SpreadElement): void {
+      const type = getConstrainedTypeAtLocation(services, node.argument);
+
+      if (isString(type)) {
+        context.report({
+          node,
+          messageId: 'noStringSpreadInArray',
+        });
+
+        return;
+      }
+    }
+
+    function checkObjectSpread(node: TSESTree.SpreadElement): void {
+      const type = getConstrainedTypeAtLocation(services, node.argument);
+
+      if (isPromise(services.program, type)) {
+        context.report({
+          node,
+          messageId: 'noSpreadInObject',
+          data: {
+            type: 'Promise',
+          },
+        });
+
+        return;
+      }
+
+      if (isFunctionWithoutProps(type)) {
+        context.report({
+          node,
+          messageId: 'noFunctionSpreadInObject',
+        });
+
+        return;
+      }
+
+      if (isIterable(type, checker) && !isString(type)) {
+        context.report({
+          node,
+          messageId: 'noSpreadInObject',
+          data: {
+            type: 'Iterable',
+          },
+        });
+
+        return;
+      }
```

Comment:
I'm just not familiar enough with custom iterables to know for sure.

Cc @typescript-eslint/triage-team

Reply:
Note: It looks like the "array in object spread" case is handled via the "iterator" handling.

If we do remove this handling or put this handling behind an option then we should special-case arrays and tuples so they're always banned, given that that's a concrete, well-defined cases that you'd never ever want, and because it's easy to catch from the type system side of things (there's utils to detect the array type)

Reply:
Here's what I know: you would almost never have a custom iterable that actually spreads and remains iterable. This is because usually `[Symbol.iterator]()` is defined on the class (such as `Set.prototype[@@iterator]`), not as an own property of each instance, and it's usually bad practice to make methods enumerable in any case.

I would suggest forbidding all iterables from being spreaded as objects, with an option that allows certain types. There's probably no reason you want to spread maps or sets but there might be reasons to spread other iterables.